velox/functions/sparksql/tests/ToJsonTest.cpp

```diff
@@ -0,0 +1,250 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <functions/sparksql/tests/SparkFunctionBaseTest.h>
+
+using namespace facebook::velox::test;
+
+namespace facebook::velox::functions::sparksql::test {
+namespace {
+constexpr float kNaNFloat = std::numeric_limits<float>::quiet_NaN();
+constexpr float kInfFloat = std::numeric_limits<float>::infinity();
+constexpr double kNaNDouble = std::numeric_limits<double>::quiet_NaN();
+constexpr double kInfDouble = std::numeric_limits<double>::infinity();
+
+class ToJsonTest : public SparkFunctionBaseTest {
+ protected:
+  core::CallTypedExprPtr createToJson(const TypePtr& inputType) {
+    std::vector<core::TypedExprPtr> inputs = {
+        std::make_shared<core::FieldAccessTypedExpr>(inputType, "c0")};
+    return std::make_shared<const core::CallTypedExpr>(
+        VARCHAR(), std::move(inputs), "to_json");
+  }
+
+  void testToJson(const VectorPtr& input, const VectorPtr& expected) {
+    auto expr = createToJson(input->type());
+    testEncodings(expr, {input}, expected);
+  }
+};
+
+TEST_F(ToJsonTest, basicStruct) {
+  auto input = makeRowVector(
+      {"a", "b"},
+      {makeNullableFlatVector<int64_t>({1, std::nullopt, 3, std::nullopt}),
+       makeNullableFlatVector<double>({1.1, 2.2, std::nullopt, std::nullopt})});
+  auto expected = makeFlatVector<std::string>(
+      {R"({"a":1,"b":1.1})",
+       R"({"a":null,"b":2.2})",
+       R"({"a":3,"b":null})",
+       R"({"a":null,"b":null})"});
+  testToJson(input, expected);
+}
+
+TEST_F(ToJsonTest, basicArray) {
+  auto input = makeNullableArrayVector<int64_t>(
+      {{{1}}, {{2, std::nullopt, 3}}, {{}}, std::nullopt});
+  auto expected = makeNullableFlatVector<std::string>(
+      {R"([1])", R"([2,null,3])", R"([])", std::nullopt});
+  testToJson(input, expected);
+}
+
+TEST_F(ToJsonTest, basicMap) {
+  auto input = makeNullableMapVector<std::string, int64_t>(
+      {{{{"a", 1}, {"b", 2}, {"c", 3}}},
+       std::nullopt,
+       {{}},
+       {{{"a", 1}, {"b", std::nullopt}}}});
+  auto expected = makeNullableFlatVector<std::string>(
+      {R"({"a":1,"b":2,"c":3})", std::nullopt, R"({})", R"({"a":1,"b":null})"});
+  testToJson(input, expected);
+}
+
+TEST_F(ToJsonTest, basicBool) {
+  auto data = makeNullableFlatVector<bool>({true, false, std::nullopt});
+  auto input = makeRowVector({"a"}, {data});
+  auto expected = makeFlatVector<std::string>(
+      {R"({"a":true})", R"({"a":false})", R"({"a":null})"});
+  testToJson(input, expected);
+}
+
+TEST_F(ToJsonTest, basicString) {
+  auto data = makeNullableFlatVector<std::string>(
+      {"str1", "str2", std::nullopt, "str\"3\"", std::nullopt});
+  auto input = makeRowVector({"a"}, {data});
+  auto expected = makeFlatVector<std::string>(
+      {R"({"a":"str1"})",
+       R"({"a":"str2"})",
+       R"({"a":null})",
+       R"({"a":"str\"3\""})",
+       R"({"a":null})"});
```

Comment:
Could you add a test case covering UTF-8 characters with lengths from 1 to 4 bytes?
```cpp
// data element:
"a\"é你😄"

// expected element:
R"({"a":"a\"é你\uD83D\uDE04"})"
```

Reply:
Good point, will add it.

Reply:
Sorry. After testing with Spark 3.5, I found that `to_json` in Spark SQL doesn't represent emojis using Unicode escape sequences. Instead, it outputs them directly in UTF-8 encoding. So the expected element should be:

```
R"({"a":"a\"é你😄"})"
```
<img width="815" alt="emoji_to_json" src="https://github.com/user-attachments/assets/d875b0b6-d900-4485-a977-6bd65c47d52a" />